### DISCUSSION TITLE
3D Product customizer

### DISCUSSION THREAD

**Discussion by aviggiano** [google/model-viewer#1820] [Show and tell]
Hi all!

Please check out our [3D product customizer](https://customizer.r2u.io/) we're building at [R2U](https://r2u.io) using model-viewer's Scene Graph API.

https://user-images.githubusercontent.com/3029017/104605387-33722d00-565d-11eb-978b-31a4d59f6581.mov

In addition to changing the material/texture, you can click "View in your space" to see the customized product in augmented reality on your mobile phone. Feedbacks are much appreciated!

I'd also like to give a big shoutout to @MayaraSs, @FelipeR2U, @daniel-limar2u and the whole @r2u-io team for this project 😃

**Comment by elalish** [comment]
Nice work! We've been working on a similar QR-code view in AR feature for our editor, which will be available imminently. We agree this is a really important piece for an editor, and we're so glad to see more web editors showing up now. Thanks for sharing!

**Reply by aviggiano** [reply]
Cool! How do you plan to provide the customized USDZ file for iOS users? We currently have an internal conversion pipeline that does that, but it would be great if it was possible to do it on the client side.

**Reply by elalish** [reply]
No plan. I'm not going to put any effort into pushing a proprietary format like USDZ. Hopefully we'll get WebXR on Safari and then we can forget about it entirely. I may link to external sites that can do it, but everything I'm aware of happens server-side anyway.

**Reply by elalish** [reply]
Hmm, well there is this: https://twitter.com/mrdoob/status/1350488054616780807

If someone can help get that exporter to stable status then we may just be able to do conversion on the fly and skip `ios-src` entirely. That would sure be nice!

**Reply by aviggiano** [reply]
Pretty sweat! I'll take a look into this. Thank you very much.

**Reply by AkshayPL** [reply]
Quite curious on the QR feature, I'm trying to make something similar as a portfolio project. Are you sending the configured 3d model back to a server for mobile users to scan the QR code and see the same model in AR?

**Reply by aviggiano** [reply]
Yes, we have a backend server that converts the 3D model between GLB/USDZ so that we can serve the same product on different devices. 

We're watching closely the [auto USDZ conversion](https://github.com/google/model-viewer/pull/2374) that has been gaining attention, and also the WebXR compatibility that will come soon, but right now this is the best solution we have come up with.

**Reply by AkshayPL** [reply]
Thats great. Also on the USDZ conversion at client side, I have made some tweaks to the USDZExporter code from Mr doob that fixes issues like opacity and hiding invisible components. Let me know, if you need the code, happy to share

**Reply by elalish** [reply]
@AkshayPL That's great, I hope you're upstreaming those fixes to three.js? I'm sure @mrdoob and all of the model-viewer users will appreciate it.

**Reply by AkshayPL** [reply]
MrDoob has created a PR for the issues, these should be merged in the next versions of Modelviewer I guess.
https://twitter.com/mrdoob/status/1394384654648684545?s=21

**Reply by elalish** [reply]
Done in v1.7!

**Comment by canberka** [comment]
This is game changer work for everybody. I really congratulations all of you. Good Job! GG!

**Comment by aviggiano** [comment]
Hi everyone, just passing by to showcase our 3D product customizer in production 😃 😃 😃 

Check it out [here](https://www.woodprime.com.br/poltrona-luis-xv-wood-prime-38725/p)! (in Portuguese)


https://user-images.githubusercontent.com/3029017/122424550-30110380-cf65-11eb-836f-7bcbd9b22445.mov



I'd like to give special thanks to @elalish and the whole model-viewer team (both from Google and also to the community) for allowing this to happen.

More to come!!!
